### PR TITLE
Make SourceClonedParameterSymbol abstract

### PR DESCRIPTION
**Pull request by Youssef1313**
~~Builds on https://github.com/dotnet/roslyn/pull/54132 (I'll rebase once the other PR goes in).~~ I think there is no need to wait on the other PR. Going to rebase on the current feature branch.

@AlekseyTs Let me know if you want the other PR to contain everything.

Proposal: dotnet/csharplang#287
Test plan: https://github.com/dotnet/roslyn/issues/52745

### CODE REVIEW

**Comment by Youssef1313** [comment]
@AlekseyTs This is ready for review. Thanks!

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:14]: >{ [](http://example.com/codeflow?start=8&length=1)

Do we want to assert that the ```originalParam``` is not optional?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:14]: @AlekseyTs I think I should be asserting `this.IsOptional`, not `originalParam.IsOptional`.

`originalParam` can be optional, but it's suppressed via `suppressOptional: true`.

I'm currently blocked on adding the assertion. See https://github.com/dotnet/roslyn/pull/54389.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:14]: > I think I should be asserting this.IsOptional

Agree

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:23]: @AlekseyTs The created parameter may not be optional as you stated in the other PR. However, I think this is still unreachable. We attempt to bind default arguments only if overload resolution succeeds.

If overload resolution succeeds, then an explicit value was passed to the parameter, meaning we are not going to access this property.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:23]: > If overload resolution succeeds, then an explicit value was passed to the parameter, meaning we are not going to access this property.

That maybe true for EndEnvoke (I assume because parameters are ref/out/in), but why this is true for BeginInvoke is not obvious to me.

> I'm open to adding any test cases you'd like if the existing tests are not enough

I think we should have targeted test for Begin/EndInvoke invocations for scenarios when cloned parameters are optional and have the attribute on them.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:23]: Both `BeginInvoke` and `EndInvoke` has extra parameter(s) at the end (`AsyncCallBack` and `object` for `BeginInvoke`, and `IAsyncResult` for `EndInvoke`). If overload resolution succeeds, then these parameters must have been supplied, meaning all the previous parameters has explicitly been passed.

I'll add some tests.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:23]: > Both `BeginInvoke` and `EndInvoke` has extra parameter(s) at the end (`AsyncCallBack` and `object` for `BeginInvoke`, and `IAsyncResult` for `EndInvoke`). If overload resolution succeeds, then these parameters must have been supplied, meaning all the previous parameters has explicitly been passed.

I see now why this is wrong.

**Comment by Youssef1313** [comment]
@AlekseyTs This is ready for another look. Thanks!

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:21]: >TestBeginInvoke [](http://example.com/codeflow?start=20&length=15)

Do we have similar tests for EndInvoke? Even though they  are likely going to be error scenarios, we want to make sure that binding doesn't fail in an unexpected way (a crash/assert, etc. )

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:53]: >_lazyCallerArgumentExpressionParameterIndex = i [](http://example.com/codeflow?start=31&length=47)

Please do not use an inline assignment when the code can easily be written without it.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:57]: >_lazyCallerArgumentExpressionParameterIndex = -1; [](http://example.com/codeflow?start=23&length=49)

Same comment.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:15]: >private int _lazyCallerArgumentExpressionParameterIndex = UninitializedArgumentExpressionParameterIndex; [](http://example.com/codeflow?start=8&length=104)

I am curios, is it worth all the trouble of recalculating the position? Could we just return -1 for now and see if anyone would ask us to support the scenarios.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:118]: >[CallerArgumentExpression(s1)] [Optional] [DefaultParameterValue(""default"")] string s2 [](http://example.com/codeflow?start=35&length=88)

The interesting scenario is when this parameter is getting cloned and the one that it references isn't and probably follows it. Is this the case in this unit-test. We want to test scenario where simply reusing the index from the original parameter will be observably wrong.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 6)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:118]: > The interesting scenario is when this parameter is getting cloned and the one that it references isn't 

For the parameter to get cloned, it must be `ref` if I understand correctly. `ref` parameters seem to have a false `IsOptional`. Hence, the test won't hit calling `CallerArgumentExpressionIndex`. This is regardless we're in error case or not.

I think the test you're asking for should hit `CallerArgumentExpressionIndex`. I'm unable to find a test case that fulfills all the following:

1. The `CallerArgumentExpression` parameter is cloned.
2. The parameter it refers to isn't.
3. `CallerArgumentExpressionIndex` gets hit.

I added a test that fulfills 1 and 2.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:118]: > For the parameter to get cloned, it must be ref if I understand correctly. ref parameters seem to have a false IsOptional. Hence, the test won't hit calling CallerArgumentExpressionIndex. This is regardless we're in error case or not.

We still can have a test attempting to do that. Trying all possible ways marking the parameter optional, attempting to ommit it at the call-site.

Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:118]: @AlekseyTs I added a test in 25dd0ff, and added two more now in 8ee88e0. Let me know if there are any other scenarios :)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:184]: >s2 [](http://example.com/codeflow?start=46&length=2)

Let's add a couple of more parameters to the end and change this arttribute to refer to the last parameter. The goal is to make the correspnding index to go out range of ordinals of parameters in EndInvoke.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:201]: >// (29,22): error CS9006: The CallerArgumentExpressionAttribute may only be applied to parameters with default values [](http://example.com/codeflow?start=16&length=117)

Technically, the parameter has the default value supplied via the DefaultParameterValue attribute. Since we do not complain about that in general, the error feels somewhat confusing. What happens if we make this a regular method and try to consume it from source and from metadata (emit definition and consume it from new compilation) without supplying the value for the parameter. What error(s) do we get? Perhaps we can reuse similar wording in this message?

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:253]: >// (28,33): error CS9006: The CallerArgumentExpressionAttribute may only be applied to parameters with default values [](http://example.com/codeflow?start=16&length=117)

Do we want to suppress this error, since we already reported (going to report) ERR_RefOutDefaultValue?

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:194]: >d.EndInvoke(ref s, null); [](http://example.com/codeflow?start=8&length=25)

Are we omitting the argument for the "interesting" parameter (parameter that was clned from the one with CallerArgumentExpression attribute)?

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:246]: >d.EndInvoke(ref s, null); [](http://example.com/codeflow?start=8&length=25)

Are we omitting the argument for the "interesting" parameter (parameter that was clned from the one with CallerArgumentExpression attribute)?

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:201]: We already get this error today with other caller infos regardless of whether this is a delegate.

```csharp
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

class Program
{
    static void D([CallerMemberName] [Optional] [DefaultParameterValue("default")] ref string s)
    {
    }
}

```

This is because the ref parameter has `IsOptional` equals false.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:302]: >d.EndInvoke(ref s, null); [](http://example.com/codeflow?start=8&length=25)

Are we omitting the argument for the "interesting" parameter (parameter that was clned from the one with CallerArgumentExpression attribute)?

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:201]: > We already get this error today with other caller infos regardless of whether this is a delegate.

I see, I guess we can live with that.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 10)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:253]: I'm not sure it's worth changing the behavior, especially that this behavior already exists with CallerMemberName and other caller infos. But I'll change if you think it's worth it :)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:194]: @AlekseyTs I'm not sure what you mean by this. The parameter that was cloned from the one with CallerArgumentExpression is passed the argument `ref s`.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:253]: > I'm not sure it's worth changing the behavior, especially that this behavior already exists with CallerMemberName and other caller infos. But I'll change if you think it's worth it

I do not have strong opinion on that. Given that twe have "preexisting condition", I can live with the current behavior.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:194]: > The parameter that was cloned from the one with CallerArgumentExpression is passed the argument ref s.

Eaxactly my point. Don't we want to try to push the compiler to supply the value for it from another parameter? In order to do that, we don't want to supply the corresponding argument explicitly, only supply other arguments, perhaps using named argument syntax.

**Comment by Youssef1313** [comment]
> .tools\msbuild\16.8.0-preview2.1\tools\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(10,5): error : (NETCORE_ENGINEERING_TELEMETRY=Restore) Failed to download package 'Microsoft.VisualStudio.Shell.Interop.14.0.DesignTime.16.10.31225.38' from 'https://pkgs.dev.azure.com/azure-public/3ccf6661-f8ce-4e8a-bb2e-eff943ddd3c7/_packaging/2a239fd0-3e21-40b0-b9d6-bc122fec7eb2/nuget/v3/flat2/microsoft.visualstudio.shell.interop.14.0.designtime/16.10.31225.38/microsoft.visualstudio.shell.interop.14.0.designtime.16.10.31225.38.nupkg'.

**Comment by Youssef1313** [comment]
Pinging @AlekseyTs and @333fred for review. Thanks!

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:26]: >-1 [](http://example.com/codeflow?start=72&length=2)

I'm unclear why this would be ok to return -1 from and the `IsCallerX` members shouldn't just return false. If we think think one of these is hittable, shouldn't they all be?

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:26]: @333fred They are all hittable. But @AlekseyTs has requested to drop support for begin and end invoke.

https://github.com/dotnet/roslyn/pull/54355#discussion_r659891931

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:26]: I see, thanks for the comment. In that case, we should update the comment to be an explicit statement of that info, rather than a question.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:68]: >supported [](http://example.com/codeflow?start=49&length=9)

Do you mean at runtime? Supported where?

**Comment by 333fred** [comment]
Done review pass (commit 11). Only a few small comments for clarifications.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:68]: @333fred No, I meant "supported by the compiler". ie, the compiler doesn't do anything for caller argument expressions for cloned begin/end invoke parameters.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:25]:
```suggestion
        // We don't currently support caller argument expression for cloned begin/end invoke.
```

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:25]: This is a "what" comment, not a "why" comment. We should include the information about how Begin/EndInvoke throw PNS exceptions and that we therefore don't do the index remapping for this case as we don't feel it's necessary.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:25]:
```suggestion
        // We don't currently support caller argument expression for cloned begin/end invoke since
        // they throw PlatformNotSupportedException at runtime and we feel it's unnecessary to support them.
```

**Review by 333fred — APPROVED** [review]
LGTM (commit 13). @dotnet/roslyn-compiler for a second review.

**Comment by Youssef1313** [comment]
@jcouv Can you review please? Thanks!

**Comment by Youssef1313** [comment]
Ping @jcouv

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:19]: I didn't understand why properties that are implemented the same way in `SourceDelegateClonedParameterSymbolForBeginAndEndInvoke` and `SourcePropertyClonedParameterSymbolForAccessors` were moved out of the base type.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs:19]: @jcouv This was requested by @AlekseyTs. The reasoning is to *force* new types to implement it. i.e, avoid future mistakes if a new type should implement it another way.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:391]: It looks like the added tests cover delegate scenarios. Do we already have scenarios for properties/indexers, since that's another scenario that uses `SourceClonedParameterSymbol`?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 13). Just a couple of questions

**Comment by jcouv** [comment]
Also it occurred to me that we'll want to think of possible interactions between CallerArgumentExpression and interpolated string handlers.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:391]: @jcouv There is a test for indexer:

https://github.com/dotnet/roslyn/blob/63531dc7c1227f2f2022aac1b4322d4aa1aac225/src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs#L1685

Let me know if you want to extend the tests with any scenarios.

Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:391]: Awesome, thanks

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 13)

**Comment by jcouv** [comment]
> possible interactions between CallerArgumentExpression and interpolated string handlers.

Clarified from chat with Fred. Here's possible scenario (may just fall out):
What if I put CallerArgumentExpression on a parameter of handler constructor, then use the argument attribute to copy an argument from invocation and pass it to that constructor?

**Pull request merged**
